React-ified AEM core components for Give site

j2trumpet opened the pull request:
There have been some instances where the documentation's demo markup doesn't match what I see actually being output in our AEM installations, so I've used the GitHub repo as the final source of truth.

[AEM Core Component Documentation](https://www.aemcomponents.dev/content/core-components-examples/library.html)

[AEM Core Component GitHub](https://github.com/adobe/aem-core-wcm-components/tree/main/content/src/content/jcr_root/apps/core/wcm/components)

j2trumpet committed:
Starting React versions of AEM core components

List, tabs, teaser, text, and title

j2trumpet committed:
Adding separator

j2trumpet committed:
Search

j2trumpet committed:
Image

j2trumpet committed:
Image refinement

j2trumpet committed:
Separator default color update

j2trumpet committed:
Adding Material Icons font

j2trumpet committed:
Breadcrumb

j2trumpet committed:
Accordion

j2trumpet committed:
Un-nesting components

j2trumpet committed:
Component template

j2trumpet committed:
Updating classes on ComponentTemplate

j2trumpet committed:
ComponentTemplate updates

j2trumpet committed:
Starting Button component

j2trumpet committed:
Finished Buttons

j2trumpet committed:
Refining Buttons

j2trumpet committed:
Alphabetizing examples

j2trumpet committed:
Folders and CSS for reminder of components to build

j2trumpet committed:
Button refinements

j2trumpet committed:
Moving className

j2trumpet committed:
Removing attribute that's throwing error

j2trumpet committed:
Carousel

j2trumpet committed:
Form text fields

j2trumpet committed:
Removing unneeded error prop

j2trumpet committed:
Form text fields refinement

j2trumpet committed:
Starting checkbox and radio button options

j2trumpet committed:
Refining checkboxes and radio buttons

j2trumpet committed:
Image component refactor

j2trumpet committed:
Updating port to use 4000 to avoid conflicts

j2trumpet committed:
Image improvements

j2trumpet committed:
Update Image.js

j2trumpet committed:
Update Image.js

j2trumpet committed:
Bug fix

Attribute was looking for a string value but was getting a boolean

LizzyRandall commented on give/src/components/Image/Image.js:
```diff
@@ -0,0 +1,64 @@
+  if (src === "" && alt === "") return null;
+
+  return (
+    <div className={`image cru-image${classes ? " " + classes : ""}`}>
```
You can actually do the following:
```
`image cru-image${classes || ""}`

j2trumpet replied:
If I'm understanding what your suggestion would do correctly, if a class string of "some-new-class" is passed, we'd end up with `cru-imagesome-new-class`. That's why I'm using `" " + classes` to add the needed space. I know I could simply add a space into `cru-image ${classes` and call it good, but if a class string isn't passed, we'd end up with `cru-image "` and that extra space would drive me nuts.

j2trumpet replied:
I think this is a moot point with how I've set up `<ComponentWrapper />` to handle classes.

LizzyRandall commented on give/src/components/Image/Image.js:
```diff
@@ -0,0 +1,64 @@
+  popupTitle = true, // boolean
+  classes = "", // string
+}) => {
+  if (src === "" && alt === "") return null;
```
is it only if src is empty string or if null? Otherwise you could do
```
if(!src && !alt) return null;
```
I get it is defaulted, but would handle if null was passed in too for some weird reason

j2trumpet replied:
That shorthand should work. It needs to return null if either `src` or `alt` are falsey.

LizzyRandall commented on give/src/components/Image/Image.js:
```diff
@@ -0,0 +1,64 @@
+  return (
+    <div className={`image cru-image${classes ? " " + classes : ""}`}>
+      <div className="cmp-image">
+        <LinkWrapper href={href}>
```
I know this is a first pass, but what users think a link should be an what is reality usually differs. I do a wrapper around links in order to add https:// onto just in case the user just types www.facebook.com kind of thing. Here's an example of the code
```
    if (!url.includes("http") && !url.includes("https")) {
      url = "https://" + url;
    }

j2trumpet replied:
That's a good precaution. It lines up with how I typically aim to protect users from themselves.

j2trumpet replied:
@LizzyRandall I'm revisiting your comments and looking at this one, I'm actually not going to implement it. These link URLs should be generated by AEM and in case it gives me a relative URL, I don't want to accidentally screw it up by adding my own protocol.

LizzyRandall replied:
@j2trumpet Mine is manually entered by the user so it can definitely be random. AEM generating I'm guessing would have the right URL 👍

LizzyRandall commented on give/src/components/Image/Image.js:
```diff
@@ -0,0 +1,64 @@
+      <div className="cmp-image">
+        <LinkWrapper href={href}>
+          <img
+            src={src}
```
Something that is nice with material ui is if an image is broken on the Avatar component it actually handles showing an icon so there isn't the broken image default by the browser

j2trumpet replied:
That's helpful. I haven't begun to think about adding MUI into the mix of this just yet. I've just been converting the AEM components to React. I'll edit the PR's description to include a link to the AEM documentation I'm following.

LizzyRandall reviewed (commented):
@j2trumpet Reviewed image component and made a few suggestions and FYI comments. Something that will need to be figured out is the way the CSS is because the way the components are they are hardcoded classes. Nice thing about styling within a component is you can see what the css is doing directly in it, otherwise you have to go to the css file and match up class name to class name. Don't do a typo! Anyway, looks good.

j2trumpet commented:
> Something that will need to be figured out is the way the CSS is because the way the components are they are hardcoded classes.

Ideally, we'll be able to use some of the styling that's currently being created for the cru.org redesign when we rebuild the Give site with React. That's why I've hardcoded the AEM classes.

j2trumpet committed:
Added Github source and one of Lizzy's suggestions

j2trumpet committed:
Cleaning up text examples

j2trumpet committed:
Cleaning up Title component

Changed "type" prop to "component" to match how MUI defines the element to use
Added optional links
Created <DynamicHeadingTag />

j2trumpet committed:
Cleaning up Breadcrumbs component

Streamlined output conditional
Improved default color assignment in case of bad given values
Added <LinkWrapper /> to avoid duplication in <BreadcrumbItem />

j2trumpet committed:
Improving Title defaults

Ensuring bad values can't be passed

j2trumpet commented:
@LizzyRandall The text, title, and breadcrumb components are ready for review if you've got a few minutes.

LizzyRandall commented on give/src/components/Breadcrumb/Breadcrumb.js:
```diff
@@ -0,0 +1,81 @@
+}) => {
+  if (content.length === 0) return null;
+
+  const classes = className !== "" ? " " + className : className;
```
I would put the spaces only in the className on the div so you don't have them all over the place. This one makes sense but the color one is always going to be something so it doesn't need the added space

j2trumpet replied:
Good catch. https://github.com/CruGlobal/cru-content-designs/pull/8/commits/322b39ab52b0b16a46c4bffa7e76b33c8d4925b0

LizzyRandall commented on give/src/components/Breadcrumb/Breadcrumb.js:
```diff
@@ -0,0 +1,81 @@
+  colorClass += colors.includes(color) ? color : colors[1];
+
+  return (
+    <div className={"breadcrumb cru-breadcrumb" + colorClass + classes}>
```
So this would be
```
className={`breadcrumb cru-breadcrumb ${colorClass}${className}`}

j2trumpet replied:
Also converting to template literals. I'm still getting used to using those. https://github.com/CruGlobal/cru-content-designs/pull/8/commits/322b39ab52b0b16a46c4bffa7e76b33c8d4925b0

LizzyRandall commented on give/src/components/Breadcrumb/Breadcrumb.js:
```diff
@@ -0,0 +1,81 @@
+  const classes = className !== "" ? " " + className : className;
+
+  const colors = ["yellow", "gray-dark", "gray-medium", "gray-light", "white"];
+  let colorClass = " cru-breadcrumb-";
```
I think this let is unnecessary, usually I only use a let if changing a string dramatically like building something in a loop. You could make it a const without it looking too complicated.
```
const colorClass=`cru-breadcrumb-${colors.includes(color) ? color : colors[1]}`;

j2trumpet replied:
I'm not sure if I knew you could put conditional statements in a template literal. Good to know! https://github.com/CruGlobal/cru-content-designs/pull/8/commits/322b39ab52b0b16a46c4bffa7e76b33c8d4925b0

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+
+export const Text = ({
+  className = "", // string
+  constraint = "text", // string - "text" | "textarea" | "email" | "tel" | "date" | "number" | "password"
```
An enum for this would be good, then if the developer does a typo or something they wouldn't be like "Why isn't this working or why is this looking this way?"

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+  value = "", // string
+  help = "", // string
+  placeholder = false, // bool
+  required = false, // bool
```
I do a lot of this kind of stuff in the siebel project and I don't do two separate variables to define something. So like it would be 
```
required: {
message: "Lalala"
}
```
Then the code can just check if there is required, then use the message. Or let's say someone wanted to require something but didn't need a message (for whatever reason) then you can force the logic to drive what needs to get passed instead of the developer guessing.

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+  const classes = className !== "" ? " " + className : className;
+  const isTextarea = constraint === "textarea" ? true : false;
+  const placeholderText = help && placeholder === true ? help : null;
+  const props = {
```
I'd make the props be an actual interface so you're not redefining them in the component itself.

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+  disabled = false, // bool
+  id = "", // string
+}) => {
+  const classes = className !== "" ? " " + className : className;
```
I feel like with this done all over the place maybe a util method called 
```
//adds space for class attribute if it exists
getClass=(className?:string)=>{
 return className !== "" ? " " + className : className
}
```

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+        {help !== "" && placeholder === false ? (
+          <p className="cmp-form-text__help-block">{help}</p>
+        ) : null}
+        {isTextarea ? (
```
With an enum you could then do a switch (maybe in a method) in order to handle all these different things. You may want methods for each of these things like
```
const helperText = (text?:string, placeHolder?:string)=>{
          !placeholder ? <p className="cmp-form-text__help-block">{help}</p> : null
}
```
Notice I make placeholder a string instead of a boolean. If a lot of things are optional, you could make it assume if it is passed, we want it shown.

LizzyRandall commented on give/src/components/Form/Text/Text.js:
```diff
@@ -0,0 +1,68 @@
+            className="cmp-form-text__textarea"
+            rows={lines}
+            {...props}
+          ></textarea>
```
Text area in it's own method would make it more like "here is a text area" and if you have enough properties and varying on the text area maybe it should be a component itself, even if it is a small one but I'm guessing it could get used in more than one place.

LizzyRandall commented on give/src/components/Title/Title.js:
```diff
@@ -0,0 +1,70 @@
+
+import "./Title.css";
+
+const LinkWrapper = ({ href, children }) => {
```
Is this link wrapper elsewhere? Sounds like it should be its own component

j2trumpet replied:
Yeah, I was thinking that same thing. I've created some version of this in three of the four components I've overhauled recently.

j2trumpet replied:
https://github.com/CruGlobal/cru-content-designs/pull/8/commits/88e6912290a450091d57cf2aab195aaa31953570

LizzyRandall commented on give/src/components/Title/Title.js:
```diff
@@ -0,0 +1,70 @@
+  const DynamicHeadingTag = tags.includes(component) ? `${component}` : "h2";
+
+  const sizes = ["xxl", "xl", "lg", "md", "sm", "xs", "xxs"];
+  let sizeClass = " cru-title-";
```
Could be a const too

j2trumpet replied:
https://github.com/CruGlobal/cru-content-designs/pull/8/commits/b4de72764f0f9492332544f480904f73a3e071fe

LizzyRandall commented on give/src/components/Title/Title.js:
```diff
@@ -0,0 +1,70 @@
+
+export const Title = ({
+  className = "", // string
+  component = "", // string = "h1" | "h2" | "h3" | "h4" | "h5" | "h6"
```
Component seems like a weird name for this. When I goggle something like scaleCategory or category might be a way to define it.

j2trumpet replied:
I originally called that prop `type` but changed it to `component` because that's what [MUI calls it](https://mui.com/material-ui/api/typography/#props) on their `<Typography />` component (and others). I figured I'd follow a convention created by (I assume) more experienced developers than I. I really don't have a strong personal opinion one way or the other.

LizzyRandall replied:
@j2trumpet It actually looks like from the docs it would be called variant
```

variant | 'body1'\| 'body2'\| 'button'\| 'caption'\| 'h1'\| 'h2'\| 'h3'\| 'h4'\| 'h5'\| 'h6'\| 'inherit'\| 'overline'\| 'subtitle1'\| 'subtitle2'\| string
-- | --


```

j2trumpet replied:
The `variant` prop sets the style, but the `component` prop sets the element. So `<Typography component="h1" variant="h2">...</Typography>` would render an `<h1>...</h1>` that's styled like the default `<h2 />`. I believe I'm doing something like this on my personal preferences screen for MPDx.

However, for this particular project, I just renamed my prop `element` to be explicit about the type of content it's expecting.

j2trumpet replied:
https://github.com/CruGlobal/cru-content-designs/pull/8/commits/88e6912290a450091d57cf2aab195aaa31953570

LizzyRandall commented on give/src/components/Title/Title.js:
```diff
@@ -0,0 +1,70 @@
+export const Title = ({
+  className = "", // string
+  component = "", // string = "h1" | "h2" | "h3" | "h4" | "h5" | "h6"
+  size = "", // string - "xxl" | "xl" | "lg" | "md" | "sm" | "xs" | "xxs"
```
enum might be good for this with a mistype. I would have to go to this component to understand any of these parameters.
Did Christian tell you about the awesome comments? They go like this:
/** Comment here*/ - so when a developer starts to use a component these kind of things like lots of parameters can be described here. Having an interface for the props is nice too so that you can hover over and see what props are needed and typescript can scream at you if something is required.

j2trumpet replied:
Isn't `enum` a TypeScript thing? I'm not ready to convert these just yet. Looks like I might be able to mimic `enum`s [like this](https://blog.bitsrc.io/explaining-enums-enumerations-in-plain-javascript-895a226622e3#a181), though I'm still not certain how they work.

LizzyRandall replied:
@j2trumpet enum isn't typescript, it's javascript. here's an example:
```
export const TextSize {
 "xl" = "xl",
"lg" = "lg"
}
```
Enums are good at defining stuff where you might want strings but in the case of typescript it helps force the actual string. In reality TextSize.xl = "xl" but the code can just reference the enum instead so the developer doesn't have to care what the string is. The enum string value doesn't have to match either, like I could do
```
export const TextSize {
 "xl" = "extra large yeah I like long strings",
"lg" = "lg"
}
```
Instead, but it makes it so when you are comparing oranges they are truly oranges as well. 🟠  It's a good way to make it if a string happened to need changing, it wouldn't break the code to do that too.

LizzyRandall reviewed (commented):
@j2trumpet Added my two cents on the Title, breadcrumb and text. I have some experience with these kinds of things since I made a generic text component similar to what you're trying. It can certainly explode and grow to be more with more exceptions added to the mix 😄

j2trumpet committed:
Lizzy's recommended Breadcrumb edits

j2trumpet committed:
Some of Lizzy's recommended Title edits

LizzyRandall commented on give/src/components/Text/Text.js:
```diff
@@ -0,0 +1,108 @@
+  if (!children) return null;
+
+  const classes = className !== "" ? " " + className : className;
+  const smallClass = small ? " cru-text-sm" : "";
```
What determines small? Is there only small versus normal? I get there's a title but is there any other stuff in between or does Title cover everything besides paragraph?

j2trumpet replied:
The `small` prop is for the leftover styling from the original new cru.org designs, but that'll change with the new new designs being made.

j2trumpet committed:
Improving Title component

Changed "component" prop to "element" for clarity
Created and used two new shared components
Swapped hard-coded tag for array cell value
Cleaned up extra spaces in class name lists

j2trumpet committed:
Adding use of shared components

j2trumpet committed:
Adding React Router for easy component switching

j2trumpet committed:
Teaser overhaul

Includes adding "hideWrapper" prop for ComponentWrapper
Includes inline CSS for React Router nav

j2trumpet committed:
Deleting old commented-out code

j2trumpet commented:
@LizzyRandall I'm still looking into the `enum` stuff, but the [Text component](https://github.com/CruGlobal/cru-content-designs/blob/give/give/src/components/Text/Text.js) and [Teaser component](https://github.com/CruGlobal/cru-content-designs/blob/give/give/src/components/Teaser/Teaser.js) are ready for a review.

j2trumpet committed:
Teaser logic hotfix

LizzyRandall commented on give/src/components/Teaser/Teaser.js:
```diff
@@ -0,0 +1,114 @@
+  img = { src: "", alt: "" }, // object with `src` and `alt` keys, each key containing a string
+  pre = "", // string
+  desc = "", // string
+  btns = [], // array of objects with `href` and `text` keys, each key containing a string
```
This is where an interface and putting types would be helpful for developer using this component I think:
```
export interface TeaserProps {
image:ImageProps
buttons:Array<ButtonProps>
}
```
And spelling out things rather than the shortened version

LizzyRandall commented on give/src/components/Teaser/Teaser.js:
```diff
@@ -0,0 +1,114 @@
+
+export const Teaser = ({
+  className = "", // string
+  type = "", // string - "featured" | "layered"
```
Type could benefit from enum too

LizzyRandall commented on give/src/components/Teaser/Teaser.js:
```diff
@@ -0,0 +1,114 @@
+          {/* Buttons */}
+          {btns.length > 0 && (
+            <div className="cmp-teaser__action-container">
+              {btns.map((btn) => (
```
The type would be nice in this map too so someone could reference easier to know what is needed for a button

LizzyRandall commented on give/src/components/shared.js:
```diff
@@ -0,0 +1,34 @@
+
+  const classes = `${type} cru-${type} ${className}`;
+
+  const DynamicInnerTag = type === "breadcrumb" ? "nav" : "div";
```
I know you're working through the enums, could be beneficial here

LizzyRandall reviewed (commented):
There's no typescript in this project right? I realized that after making the comments. Anyway, added more two cents 😄

j2trumpet committed:
Separator overhaul

j2trumpet committed:
Tabs overhaul

j2trumpet committed:
List overhaul

j2trumpet committed:
Adding `unwrap` property to better match display logic for anchors in AEM components

j2trumpet committed:
Button overhaul

j2trumpet committed:
Accordion overhaul

j2trumpet committed:
Carousel overhaul

j2trumpet committed:
Component template overhaul

j2trumpet commented:
@LizzyRandall I've got a few new components for you to look at. I've left the enum stuff alone for now.

LizzyRandall commented on give/src/components/Tabs/Tabs.js:
```diff
@@ -7,49 +12,55 @@ export const Tabs = ({
+      className={`${dot ? "cru-tabs-dot " : ""}${className}`}
+    >
+      <ol className="cmp-tabs__tablist">
+        {content.map((tab, index) => {
```
It's interesting to loop through the same array, just different objects in it. I think if the tabs and panels were in two separate arrays it would be clearer what is being looped through.

j2trumpet replied:
In the end, the content will be provided through AEM, so this will undoubtedly have to change. For now, I opted for the current solution so that in my `content` variable, I could neatly keep tab labels and tab content grouped together.

LizzyRandall commented on give/src/components/List/List.js:
```diff
@@ -1,71 +1,56 @@
-          {item.desc}
-        </ListItemText>
-      </article>
+      {!teaser && (
```
Might be a little cleaner to put the non teaser in a method and teaser in a method. Then it could be like
{teaser ? teaserMethod(item) : listItem(item)}

j2trumpet replied:
How's this?

```
const ListItems = ({ items, linked, teaser }) => {
  if (items.length === 0) return null;

  const listItemStandard = (item, linked) => (
    <>
      <LinkWrapper
        href={item.href}
        unwrap={!item.href || !linked}
        className="cmp-list__item-link"
      >
        <span className="cmp-list__item-title">{item.title}</span>
        <span className="cmp-list__item-date">{item.date}</span>
      </LinkWrapper>
      <span className="cmp-list__item-description">{item.desc}</span>
    </>
  );

  const listItemTeaser = (item, linked) => (
    <Teaser
      title={item.title}
      desc={item.desc}
      link={linked && item.href}
      hideWrapper
      fromList
    />
  );

  return items.map((item) => (
    <li className="cmp-list__item" key={item.title}>
      {teaser ? listItemTeaser(item, linked) : listItemStandard(item, linked)}
    </li>
  ));
};
```

LizzyRandall commented on give/src/components/Button/Button.js:
```diff
@@ -1,12 +1,19 @@
 
 const ButtonLink = ({ href, children }) => {
-  return href !== "" ? (
+  return href ? (
```
So a button will never link to a real link? I'm just curious for things like edit my designation page or something on the give site, I'm assuming that actually goes to some URL potentially? This would force it to be a link right unless the children passed is a button itself.

j2trumpet replied:
If a button AEM component is provided a valid URL, it will render as an `<a>`. Otherwise, it'll render as a `<button>`. Compare the standard and linked versions' markup from the [documentation](https://www.aemcomponents.dev/content/core-components-examples/library/core-content/button.html).

LizzyRandall replied:
@j2trumpet Ahhh and make it look like a button, got it. I was thinking it only forced links which would be weird.

LizzyRandall commented on give/src/components/Accordion/Accordion.js:
```diff
@@ -29,20 +13,36 @@ export const Accordion = ({
+          </h3>
+          <div
+            className={`cmp-accordion__panel cmp-accordion__panel--${
+              index === 0 ? "expanded" : "hidden"
```
So this forces only the first to ever be expanded, could there be a case where something not the first could be opened initially? And what about the state for the accordion to actually keep the expanded or not expanded, are these just static?

j2trumpet replied:
This is another instance where this code will need to change once we start pulling in real AEM content. That content will determine which, if any, accordion is open initially.

As for state, I'm unsure right now whether we'd want AEM or React to control the functionality and state of the accordion. This is all static for now.

LizzyRandall commented on give/src/components/Carousel/Carousel.js:
```diff
@@ -10,57 +16,66 @@ export const Carousel = ({
+            >
               <span className="cmp-carousel__action-icon"></span>
-              <span className="cmp-carousel__action-text">Pause</span>
+              <span className="cmp-carousel__action-text">${action.label}</span>
```
Wouldn't this just be {action.label} not ${} since it isn't in a template literal?

j2trumpet replied:
Good catch.

LizzyRandall commented on give/src/components/Carousel/Carousel.js:
```diff
@@ -10,57 +16,66 @@ export const Carousel = ({
-        </div>
+          return (
+            <div className={`cmp-carousel__item${active}`} key={index}>
+              {slide.slide}
```
slide.slide seems a little weird, maybe like slide.content since it's the content in the slide right?

LizzyRandall commented on give/src/components/Carousel/Carousel.js:
```diff
@@ -71,12 +86,12 @@ export const CarouselExamples = () => {
           },
           {
-            slide: <Image img={{ src: epcot, alt: "EPCOT" }} />,
+            slide: <Image src={lava2} alt="Lava 2" />,
```
Lava instead of Epcot?! Guess that makes sense 🗻

j2trumpet replied:
Lol! I'm switching out the demo images to match what's in the AEM component documentation.

LizzyRandall commented on give/src/App.js:
```diff
@@ -70,17 +70,16 @@ const Nav = () => {
-};
+// const Home = () => {
+//   return <p>Hello world!</p>;
+// };
```
Are you planning on coming back to this at some point? Or was it just to show something?

j2trumpet replied:
I forgot to delete it once I verified what I was replacing it with worked.

LizzyRandall commented on give/src/components/ComponentTemplate.js:
```diff
@@ -1,24 +1,28 @@
+        <p>
+          This is a simple component to be used as a template for building out
+          larger components.
+        </p>
```
Is this only a wrapper? Can't really tell how it is a template?

j2trumpet replied:
This is a template in the sense that it gives me a starting point for how to organize my code when building out new components. I try to be as consistent in my code as possible and this helps. It won't actually be used in the final project. Maybe I should call it something different.

LizzyRandall replied:
@j2trumpet Ah so something like bolierplate?

j2trumpet replied:
@LizzyRandall Something like that.

LizzyRandall reviewed (commented):
@j2trumpet Reviewed the last commitments and commented on the components. Looks good overall, just had a few questions and suggestions.

j2trumpet committed:
Updates from Lizzy's feedback

j2trumpet committed:
Search overhaul

Plus change from ternary to switch/case in <ComponentWrapper>

j2trumpet committed:
Preparing for form overhaul

j2trumpet committed:
Syncing Search with GitHub code

j2trumpet committed:
Demo navigation overhaul

j2trumpet committed:
Form text overhaul

j2trumpet committed:
Fixing missing key issue

j2trumpet committed:
Form options overhaul and completion

j2trumpet committed:
Form button

j2trumpet commented:
@LizzyRandall The search and form components are ready for a review.

LizzyRandall commented on give/src/components/Search/Search.js:
```diff
@@ -0,0 +1,39 @@
+  className = "", // string
+  dark = false, // bool
+}) => {
+  const darkClass = dark ? "cru-search-dark" : "";
```
Didn't you do something for classes since they all did this? It's been a while so I can't remember 😅

LizzyRandall commented on give/src/components/Form/Form.js:
```diff
@@ -0,0 +1,74 @@
+  if (!children) return null;
+
+  const underlineClass = underline ? "cru-form--underlined" : "";
+  const darkClass = dark ? "cru-form--dark" : "";
```
Can someone have a dark version of the AEM editor or something?

LizzyRandall commented on give/src/components/Form/Form.js:
```diff
@@ -0,0 +1,74 @@
+      </Form>
+      <Form underline dark>
+        <TextExamples />
+      </Form> */}
```
Is the hope to have more later? Is that why these are commented out?

LizzyRandall reviewed (commented):
@j2trumpet Looked at Form.js and Search.js!

LizzyRandall commented on give/src/components/Form/Button/Button.js:
```diff
@@ -0,0 +1,41 @@
+        type={type}
+        id={id ? id : null}
+        className="cmp-form-button"
+        name={name ? name : null}
```
Why would null be a valid option? Wouldn't "" be instead or do you not want the attribute at all in that case? I would think a generic name like "cru-button" or something could be a default for at least the name

j2trumpet replied:
I'm mimicking what AEM does and if the attributes aren't provided values, they aren't output at all. At least that's how our current AEM installation works. I don't know if this will change once we're on the cloud.

LizzyRandall commented on give/src/components/Form/Options/Options.js:
```diff
@@ -0,0 +1,172 @@
+  const isCheckbox = type === "checkbox";
+  const isRadio = type === "radio";
+  const isSelect = type === "drop-down";
+  const isMultiSelect = type === "multi-drop-down";
```
You were doing enums later right? Enum would be good here so less mistypes

LizzyRandall commented on give/src/components/Form/Options/Options.js:
```diff
@@ -0,0 +1,172 @@
+
+  const selectValues = () => {
+    let values = [];
+    options.forEach((option) => {
```
You could actually use flatMap here instead so like
```
const values = options.flatMap((option)=>{
return option.selected ? option.value : [];
});
```

j2trumpet replied:
I didn't know `flatMap()` was a thing.

https://github.com/CruGlobal/cru-content-designs/pull/8/commits/23075605aba1cf5f5cb0893c8fe53c0f116139df

LizzyRandall commented on give/src/components/Form/Options/Options.js:
```diff
@@ -0,0 +1,172 @@
+        values.push(option.value);
+      }
+    });
+    if (isMultiSelect) {
```
You can use return early and pop too for this
```
if(!isMultiSelect){
values.pop();
return values ;
}
return values;
```

j2trumpet replied:
I forgot about `.pop()`. I simplified it even more I feel.

https://github.com/CruGlobal/cru-content-designs/pull/8/commits/23075605aba1cf5f5cb0893c8fe53c0f116139df

LizzyRandall reviewed (commented):
@j2trumpet Suggested a few things on button.js, options.js

j2trumpet committed:
Lizzy's options recommendations

j2trumpet committed:
Fixing console errors

j2trumpet committed:
Fixing more console errors

j2trumpet committed:
Accordion overhaul #2

Took out links and dark version styling
Added ID props
Created separate Accordion and AccordionItem components
Took out default content array

j2trumpet commented:
@LizzyRandall New accordion overhaul.

LizzyRandall reviewed (commented):
@j2trumpet Looks good for accordion

j2trumpet committed:
Breadcrumb overhaul 2

Took out colored version styling
Added ID props
Created separate Breadcrumb and BreadcrumbItem components
Took out default content array

j2trumpet commented:
@LizzyRandall New breadcrumb overhaul.

LizzyRandall reviewed (commented):
@j2trumpet Looks good to me

j2trumpet committed:
Button overhaul 2

Plus removing default empty className for Breadcrumb

j2trumpet committed:
Carousel overhaul 2
